packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -232,10 +233,125 @@ export default createRule<Options, MessageId>({
+            // leaving us with just the `asserts x` cases.
+            returnTypeAnnotation.type == null &&
+            // I think this is redundant but, still, it needs to be true
+            returnTypeAnnotation.assertsModifier != null
+          )
+        ) {
+          return;
+        }
+
+        const assertionTarget = returnTypeAnnotation.parameterName;
+        if (assertionTarget.kind !== ts.SyntaxKind.Identifier) {
+          // This can happen when asserting on `this`. Ignore!
+          return;
+        }
+
+        const firstParameter = declaration.parameters.at(0);
+        const nonThisParameters =
+          firstParameter != null &&
+          firstParameter.name.kind === ts.SyntaxKind.Identifier &&
+          firstParameter.name.text === 'this'
+            ? declaration.parameters.slice(1)
+            : declaration.parameters;
+
+        // Don't even bother inspecting parameters past the number of
+        // arguments we have at the call site.
+        const checkableNonThisParameters = nonThisParameters.slice(
+          0,
+          checkableArguments.length,
+        );
+
+        let assertedParameterIndexForThisSignature: number | undefined;
+        for (const [index, parameter] of checkableNonThisParameters.entries()) {
+          if (parameter.dotDotDotToken != null) {
+            // Cannot assert a rest parameter, and can't have a rest parameter
+            // before the asserted parameter. It's not only a TS error, it's
+            // not something we can logically make sense of, so give up here.
+            return;
+          }
+
+          if (parameter.name.kind !== ts.SyntaxKind.Identifier) {
+            // Only identifiers are valid for assertion targets, so skip over
+            // anything like `{ destructuring: parameter }: T`
+            continue;
+          }
+
+          // we've found a match between the "target"s in
+          // `function asserts(target: T): asserts target;`
+          if (parameter.name.text === assertionTarget.text) {
+            assertedParameterIndexForThisSignature = index;
+            break;
+          }
+        }
+
+        if (assertedParameterIndexForThisSignature == null) {
+          // Didn't find an assertion target in this signature that could match
+          // the call site.
+          return;
+        }
+
+        if (
+          assertedParameterIndex != null &&
+          assertedParameterIndex !== assertedParameterIndexForThisSignature
+        ) {
+          // Found asserted parameter didn't match previous signatures.
+          return;
+        }
+
+        assertedParameterIndex = assertedParameterIndexForThisSignature;
+      }
+
+      // Didn't find a unique assertion index.
+      if (assertedParameterIndex == null) {
+        return;
+      }
+
+      const assertedParameter = checkableArguments[assertedParameterIndex];
+      traverseNode(assertedParameter, true);
```

Comment:
[Refactor] Totally optional, but I had a bit of a hard time figuring out the intended flow of all the added code. There's a lot. Extracting out the _"what is the asserted parameter index?"_ logic into a separate might be helpful in making that more clear. What do you think?

```suggestion
      const assertedParameter = getAssertedParameter(checkableArguments, node);

      if (assertedParameter) {
        traverseNode(assertedParameter, true);
      }
```

Reply:
Very much agree!

Reply:
Extracted to a local function for now.... We'll probably want to extract this function to a separate file whenever we add this feature to no-unnecessary-condition as well, see https://github.com/typescript-eslint/typescript-eslint/issues/9076